composer/trainer/trainer.py

```diff
@@ -2848,7 +2852,20 @@ def _train_microbatches(
                         optimizer.zero_grad()
 
             # Tracker for gradient accumulation
-            current_batch_size = sum([self._train_data_spec.get_num_samples_in_batch(batch) for batch in microbatches])
+            if self.accumulate_train_batch_on_tokens:
+                current_batch_size = sum([self._train_data_spec.get_num_tokens_in_batch(b) for b in microbatches])
+                if current_batch_size == 0:
+                    warnings.warn(
+                        'Requested loss accumulation based on number of tokens in training batch, but zero tokens '
+                        'found (perhaps due to an improper DataSpec).',
+                    )
+            else:
+                current_batch_size = sum([self._train_data_spec.get_num_samples_in_batch(b) for b in microbatches])
+            # Average the current batch size across ranks, to ensure each rank contributes appropriately
```

Comment:
Ok I think this is not quite correct. If we want to weight each token equally, then we want to scale the microbatch loss as
```
microbatch_loss = microbatch_loss * (microbatch_num_tokens/global_batch_size_tokens)*world_size
```

The current implementation won't be exactly this because its just scaling by the average batch size in tokens. 

Also I would prefer that we leave the scaling code the same for the original samples case, since there is no reason to adjust it.

Let me know if this makes sense or what I am missing....

Reply:
I'm pretty sure that the two are equivalent: we can write
```
  microbatch_loss * (microbatch_num_tokens / global_batch_size) * world_size
= microbatch_loss * microbatch_num_tokens / (global_batch_size / world_size)
= microbatch_loss * microbatch_num_tokens / all_reduce(current_batch_size, "AVG")
```
I'd be happy to revert for the original samples case, though I do think the updated version is more correct in the event that different ranks have different numbers of samples.

Reply:
Ah sorry, I didn't put enough parens, trying again.
```
microbatch_loss = (microbatch_loss * (microbatch_num_tokens/global_batch_size_tokens))*world_size
```

Reply:
Not sure if I'm missing something, but does that make a difference? Since it's all multiplications and divisions. The current version is `microbatch_loss * local_batch_size`, which is equivalent only if `local_batch_size` is the same on all all ranks.

For what it's worth, I do find that the version in the code here gives the same result as a `torchmetrics.Metric` I implemented (in our own repo using `composer`) which accumulates the number of tokens and the loss separately, before sum-reducing both and dividing.

Reply:
Ah sorry, found my error, you are correct. @mvpatel2000 what do you think about changing the behavior for the sample weighted case? I agree it is more correct, but its also a pretty subtle change in behavior.

Reply:
I am for changing. It's more correct -- the distributed case should match the non distributed case, and this fixes it in the scenario for uneven samples.